tests/lib.rs

```diff
@@ -111,3 +116,175 @@ fn mts_more_nodes() {
     assert_eq!(expected_t, t.hash());
   }
 }
+
+fn build_mts(
+  data: &[Vec<u8>],
+) -> (MerkleTreeStream<S, DefaultNode>, Vec<Rc<DefaultNode>>) {
+  let roots = Vec::new();
+  let mut mts = MerkleTreeStream::new(S, roots);
+  let mut nodes: Vec<Rc<DefaultNode>> = Vec::new();
+
+  data.iter().for_each(|bs| mts.next(&bs, &mut nodes));
+  (mts, nodes)
+}
+
+fn all_children(index: usize) -> Box<Iterator<Item = usize>> {
```

Comment:
can we use `impl Iterator` here?

Reply:
Not without making the body a bit less readable. The problem is that the two arms of the match have different types which is not allowed for impl Traits.

This is how it would look like:

```rust
fn all_children(index: usize) -> impl Iterator<Item = usize> {
  let self_ = iter::once(index);
  let children: Box<Iterator<Item = usize>> = match flat_tree::children(index) {
    Some((left, right)) => {
      Box::new(all_children(left).chain(all_children(right)))
    }
    None => Box::new(iter::empty()),
  };
  self_.chain(children)
}
```

Actually, I'm quite on the fence on which one is more readable.

Reply:
Ahh yeah, I see. Let's just take the path of least resistance and keep what you wrote (: